### PR TITLE
[Master] Add new option in json conversions

### LINKED ISSUE #42453: [Improvement]: Support new options in JSON conversions
### Description

$title.

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by mindula**
## Purpose
$title

Fixes #42453

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/json-to-record-converter/src/main/java/io/ballerina/jsonmapper/JsonToRecordMapper.java [hunk 1]
```diff
@@ -409,6 +412,8 @@ private static void updateRecordFields(JsonObject jsonObject, Map<String, JsonEl
                         (RecordFieldNode) getRecordField(jsonEntry, existingFieldNames, updatedFieldNames, isOptional);
                 recordField = recordField.modify()
                         .withTypeName(nonAnyDataNode == null ? unionTypeDescNode : nonAnyDataNode).apply();
+                recordField = recordField.modify()
+                        .withFieldName(optionalFieldName == null ? recordField.fieldName() : optionalFieldName).apply();
```
Comment (nipunayf) [inline misc/json-to-record-converter/src/main/java/io/ballerina/jsonmapper/JsonToRecordMapper.java:416]:
```suggestion
                recordField = recordField.modify()
                        .withTypeName(nonAnyDataNode == null ? unionTypeDescNode : nonAnyDataNode)
                        .withFieldName(optionalFieldName == null ? recordField.fieldName() : optionalFieldName)
                        .apply();
```

File: misc/json-to-record-converter/src/main/java/io/ballerina/jsonmapper/JsonToRecordMapper.java [hunk 2]
```diff
@@ -392,11 +393,13 @@ private static void updateRecordFields(JsonObject jsonObject, Map<String, JsonEl
...
+                        optionalFieldName = AbstractNodeFactory.createIdentifierToken(entry.getKey() +
+                                questionMarkToken.text());
                     } else if (node2.kind().equals(SyntaxKind.ANYDATA_KEYWORD)) {
-                        nonAnyDataNode =
-                                NodeParser.parseTypeDescriptor(node1.toSourceCode() + questionMarkToken.text());
+                        nonAnyDataNode = NodeParser.parseTypeDescriptor(node1.toSourceCode());
+                        optionalFieldName = AbstractNodeFactory.createIdentifierToken(entry.getKey() +
+                                questionMarkToken.text());
                     }
                 }
```
Comment (nipunayf) [inline misc/json-to-record-converter/src/main/java/io/ballerina/jsonmapper/JsonToRecordMapper.java:404]: We can rewrite this logic as below given that node.kind() is extracted to a variable.

```java
else if (node1Kind.equals(SyntaxKind.ANYDATA_KEYWORD) ||
                        node2Kind.equals(SyntaxKind.ANYDATA_KEYWORD)) {
    nonAnyDataNode = NodeParser.parseTypeDescriptor(node1Kind.equals(SyntaxKind.ANYDATA_KEYWORD) ?
            node2.toSourceCode() : node1.toSourceCode());
    Token questionMarkToken = AbstractNodeFactory.createToken(SyntaxKind.QUESTION_MARK_TOKEN);
    optionalFieldName = AbstractNodeFactory.createIdentifierToken(entry.getKey() +
            questionMarkToken.text());
}
```

Comment (mindula) [reply misc/json-to-record-converter/src/main/java/io/ballerina/jsonmapper/JsonToRecordMapper.java:404]: Let's not focus on these improvements for the moment as this PR is urgent

**Review by LakshanWeerasinghe — APPROVED** [review]

**Review by KavinduZoysa — APPROVED** [review]

**Review by LakshanWeerasinghe — DISMISSED** [review]

**Review by prakanth97 — CHANGES_REQUESTED** [review]
This has to be supported as an option instead of overwriting the existing behaviour

File: misc/json-to-record-converter/src/main/java/io/ballerina/jsonmapper/JsonToRecordMapper.java [hunk 3]
```diff
@@ -122,7 +123,7 @@ public static JsonToRecordResponse convert(String jsonString, String recordName,
      */
     public static JsonToRecordResponse convert(String jsonString, String recordName, boolean isRecordTypeDesc,
                                                boolean isClosed, boolean forceFormatRecordFields, String filePathUri,
-                                               WorkspaceManager workspaceManager) {
+                                               WorkspaceManager workspaceManager, boolean isOptionalField) {
```
Comment (prakanth97) [inline misc/json-to-record-converter/src/main/java/io/ballerina/jsonmapper/JsonToRecordMapper.java:126]: Shall we rename `isOptionalField`, `isNullAsOptional`?

**Review by prakanth97 — APPROVED** [review]
LGTM

**Review by LakshanWeerasinghe — APPROVED** [review]

**Pull request merged**